org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/CaptureBinding.java

```diff
@@ -596,7 +609,8 @@ public String toString() {
 		if (this.firstBound instanceof ArrayBinding) {
 			this.signature = constantPoolName();
 		} else {
-			this.signature = CharOperation.concat('L', constantPoolName(), ';');
+			char[] poolName = this.wildcard != null ? this.wildcard.constantPoolName() : constantPoolName();
+			this.signature = CharOperation.concat('L', poolName, ';');
```

Comment:
So, this change is needed in a situation where `CaptureBinding.firstBound == null` but the `wildcard` has a relevant upper bound.

Given "The erasure of a type variable ([§4.4](https://docs.oracle.com/javase/specs/jls/se20/html/jls-4.html#jls-4.4)) is the erasure of its leftmost bound." (JLS 4.6) we seem to be inconsistent as to what is a capture's leftmost bound. The super implementation assumes, that's either `firstBound` or `superclass`, but doesn't consider `superinterfaces`. Assuming this might be correct for `TypeVariableBinding`, perhaps `CaptureBinding.initializeBounds()` should be blamed? Why don't we set `firstBound`?

It happens in the 2nd `case Wildcard.UNBOUND`, where the following addition seems to render the change in `signature` unnecessary (line 266):

```
	if (substitutedVariableSuperclass.id == TypeIds.T_JavaLangObject && substitutedVariableInterfaces.length > 0) {
		this.setFirstBound(substitutedVariableInterfaces[0]);
	}
```

To me this looks like a fix closer to the root cause. WDYT?

Reply:
I ran the full test suite with the proposed change and see two extra failures:

1. org.eclipse.jdt.core.tests.model.NullAnnotationModelTests.testBug495635() and
2. org.eclipse.jdt.core.tests.model.CompletionTests_1_5.test0356()

The first one is an NPE at 
```
	if (substitutedVariableSuperclass.id == TypeIds.T_JavaLangObject && substitutedVariableInterfaces.length > 0) {
		this.setFirstBound(substitutedVariableInterfaces[0]);
	}
```

which I can circumvent by null checks - Although I am not (yet at least) sure if there is some thing wrong in seeing substitutedVariableSuperclass against the TypeVariable `O` in the test case.

The other test seems to fails with
```
junit.framework.ComparisonFailure: 
----------- Expected ------------
getClass[METHOD_REF]{getClass(), Ljava.lang.Object;, ()Ljava.lang.Class<+Ljava.lang.Object;>;, getClass, null, 60}\n
get[METHOD_REF]{get(), Ltest.util.List<Ljava.lang.Object;>;, (I)Ljava.lang.Object;, get, (i), 64}
------------ but was ------------
getClass[METHOD_REF]{getClass(), Ljava.lang.Object;, ()Ljava.lang.Class<+Ljava.lang.Object;>;, getClass, null, 60}\n
get[METHOD_REF]{get(), Ltest.util.List<Ltest.util.List<TT;>;>;, (I)Ltest.util.List<TT;>;, get, (i), 64}
```

Looking into both

Reply:
I'll also do some wider impact analysis of the proposed change.

Reply:
> To me this looks like a fix closer to the root cause. WDYT?

Not having any objections, I will adopt the suggested change - Thanks!